dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/converter/TrackerBundleParamsConverter.java

```diff
@@ -84,174 +82,238 @@
+                .forEach( enrollment -> enrollmentHashMap.put( enrollment.getEnrollment(), enrollment ) );
 
-    private void flattenPayload( TrackerBundleParams bundle )
-    {
+            extractRelationships( te )
+                .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
+        }
 
-        List<Event> events = new ArrayList<>();
+        // Set UID for all enrollments and notes
+        dataBundle.getEnrollments().stream()
+            .peek( enrollment -> updateEnrollmentReferences( enrollment, enrollment.getTrackedEntity() ) )
+            .forEach( enrollment -> enrollmentHashMap.put( enrollment.getEnrollment(), enrollment ) );
 
-        List<Enrollment> enrollments = new ArrayList<>();
+        // Extract all events and relationships, and set parent references
+        for ( Enrollment enrollment : enrollmentHashMap.values() )
+        {
+            extractEvents( enrollment )
+                .forEach( event -> eventHashMap.put( event.getEvent(), event ) );
 
-        Set<Relationship> relationships = new HashSet<>();
+            extractRelationships( enrollment )
+                .forEach( relationship -> relationshipHashMap.put( relationship.getRelationship(), relationship ) );
 
-        // Iterate over **all** enrollments
-        for ( Enrollment enrollment : bundle.getTrackedEntities().stream()
-            .flatMap( l -> l.getEnrollments().stream() ).collect( Collectors.toList() ) )
-        {
-            // collect all events from enrollments and add them to the flattened events collection
-            events.addAll( enrollment.getEvents().stream().map( e -> addParent( e, enrollment.getEnrollment() ) )
-                .collect( Collectors.toList() ) );
-
-            // Extract relationships from enrollment
-            relationships.addAll( enrollment.getRelationships() );
-
-            // Extract relationships from events
-            relationships
-                .addAll( events.stream().flatMap( l -> l.getRelationships().stream() ).collect( Collectors.toSet() ) );
-
-            // remove events and relationships from enrollment
-            enrollment.setEvents( Collections.emptyList() );
-            enrollment.setRelationships( Collections.emptyList() );
-            // remove relationships from events
-            events.forEach( e -> e.setRelationships( Collections.emptyList() ) );
+            enrollment.setNotes( enrollment.getNotes().stream()
+                .filter( note -> !StringUtils.isEmpty( note.getValue() ) )
+                .peek( this::updateNoteReferences )
+                .collect( Collectors.toList() )
+            );
         }
 
-        for ( TrackedEntity trackedEntity : bundle.getTrackedEntities() )
+        // Set UID for all events and notes
+        dataBundle.getEvents().stream()
+            .peek( event -> updateEventReferences( event, event.getTrackedEntity(), event.getEnrollment() ) )
```

Comment:
We are updating references only for the events that are coming as a separate entity in the payload.
If we have an event inside an enrollment and this event has no uid, at some point the validation will fail.
Is this expected?

Reply:
The references are updated within "extractEvents", so this should not be a problem :)